sklearn/_loss/tests/test_loss.py

```diff
@@ -0,0 +1,814 @@
+        y_pred[:, 1] = 0.5 * (1 - y_pred[:, 0])
+        y_pred[:, 2] = 0.5 * (1 - y_pred[:, 0])
+
+    assert loss.in_y_pred_range(y_pred)
+
+    # calculating losses should not fail
+    raw_prediction = loss.link(y_pred)
+    loss.loss(y_true=y_true, raw_prediction=raw_prediction)
+
+
+@pytest.mark.parametrize(
+    "loss, y_true_success, y_true_fail",
+    [
+        (HalfSquaredError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (AbsoluteError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (PinballLoss(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (HalfPoissonLoss(), [0, 0.1, 100], [-np.inf, -3, -0.1, np.inf]),
+        (HalfGammaLoss(), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (HalfTweedieLoss(power=-3), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (HalfTweedieLoss(power=0), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (
+            HalfTweedieLoss(power=1.5),
+            [0, 0.1, 100],
+            [-np.inf, -3, -0.1, np.inf],
+        ),
+        (HalfTweedieLoss(power=2), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (HalfTweedieLoss(power=3), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (BinaryCrossEntropy(), [0, 0.5, 1], [-np.inf, -1, 2, np.inf]),
+        (CategoricalCrossEntropy(), [0.0, 1.0, 2], [-np.inf, -1, 1.1, np.inf]),
+    ],
+)
+def test_loss_boundary_y_true(loss, y_true_success, y_true_fail):
+    # Test boundaries of y_true for loss functions.
+    for y in y_true_success:
+        assert loss.in_y_true_range(np.array([y]))
+    for y in y_true_fail:
+        assert not loss.in_y_true_range(np.array([y]))
+
+
+@pytest.mark.parametrize(
+    "loss, y_pred_success, y_pred_fail",
+    [
+        (HalfSquaredError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (AbsoluteError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (PinballLoss(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
+        (HalfPoissonLoss(), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (HalfGammaLoss(), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (
+            HalfTweedieLoss(power=-3),
+            [0.1, 100],
+            [-np.inf, -3, -0.1, 0, np.inf],
+        ),
+        (HalfTweedieLoss(power=0), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (
+            HalfTweedieLoss(power=1.5),
+            [0.1, 100],
+            [-np.inf, -3, -0.1, 0, np.inf],
+        ),
+        (HalfTweedieLoss(power=2), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (HalfTweedieLoss(power=3), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
+        (BinaryCrossEntropy(), [0.1, 0.5], [-np.inf, 0, 1, np.inf]),
+        (CategoricalCrossEntropy(), [0.1, 0.5], [-np.inf, 0, 1, np.inf]),
+    ],
+)
+def test_loss_boundary_y_pred(loss, y_pred_success, y_pred_fail):
+    # Test boundaries of y_pred for loss functions.
+    for y in y_pred_success:
+        assert loss.in_y_pred_range(np.array([y]))
+    for y in y_pred_fail:
+        assert not loss.in_y_pred_range(np.array([y]))
```

Comment:
Because `y_pred` and `y_true` do not always have the same domain (valid value range). Good examples are
- Binomial/BinCrosEntr: while `0 <= y_true <= 1`, one has `0 < y_pred < 1`.
- Poisson: `0 <= y_true`, but `0 < y_pred`.

In both cases, the corresponding predictive distribution with `y_pred = 0` would be a point mass at zero.

Reply:
OK, thanks for the explanation.

For maintainability, clarity and extendibility purposes, probably we could create at the head of the file 2 tuples those tests from a common tuple (containing commons set of parameters) and include your explanation in a small surrounding comment.

Reply:
Could you sketch what you've suggested? Either I don't get it or I don't see how it results in better maintainability:man_shrugging:

Reply:
Yes, I could have been clearer.

Here, some parameters values are used for both tests.
This is for instance the case for those parameters, which are here grouped under `COMMON_PARAMS`:
```python
# Fixture to test valid value ranges.
COMMON_PARAMS = [
        (HalfSquaredError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
        (AbsoluteError(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
        (PinballLoss(), [-100, 0, 0.1, 100], [-np.inf, np.inf]),
        (HalfPoissonLoss(), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
        (HalfGammaLoss(), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
        (HalfTweedieLoss(power=2), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
        (HalfTweedieLoss(power=3), [0.1, 100], [-np.inf, -3, -0.1, 0, np.inf]),
]
```

What I suggest, it to create for each test another iterable of parameters values used for this test solely.
For instance, with parts of your previous answer as a comment:
```python
# y_pred and y_true do not always have the same domain (valid value range).
# Hence we define extra set of parameters for each.
Y_TRUE_PARAMS = [
        (BinaryCrossEntropy(), [0, 0.5, 1], [-np.inf, -1, 2, np.inf]),
        (CategoricalCrossEntropy(), [0.0, 1.0, 2], [-np.inf, -1, 1.1, np.inf]),
        # ...
]

Y_PRED_PARAMS = [
        # ... similarly
]
```

So that we factor them in the parametrisations' definitions:

```python
@pytest.mark.parametrize(
    "loss, y_true_success, y_true_fail", [*COMMON_PARAMS, *Y_TRUE_PARAMS]
)
def test_loss_boundary_y_true(loss, y_true_success, y_true_fail):
    # Test boundaries of y_true for loss functions.
    #...

#...

@pytest.mark.parametrize(
    "loss, y_true_success, y_true_fail", [*COMMON_PARAMS, *Y_PRED_PARAMS]
)
def test_loss_boundary_y_pred(loss, y_pred_success, y_pred_fail):
    # Test boundaries of y_pred for loss functions.
    # ...
```

`COMMON_PARAMS`, `Y_TRUE_PARAMS`, and `Y_PRED_PARAMS`  could be set at the head of the file as done for tests' fixtures generally (I am referring to [such fixtures](https://github.com/scikit-learn/scikit-learn/blob/8c8ea9e19f37a0ba35be11240acc68ea80f7f431/sklearn/neighbors/tests/test_neighbors_tree.py#L27-L38) for instance) and probably their name could be changed for something clearer (?).

Let me know if I am still not clear. 🙂

Reply:
Hi @lorentzenchr, 

Does this makes my initial comment clearer? 🙂

Reply:
Much clearer. I gave it a try.